### ISSUE TITLE
Document Grafana integration (adding dashboards, data persistence)

### ISSUE THREAD

**Issue by zytek** [coreos/prometheus-operator#1326]
As a (first time) user of Prometheus stack I'm finding it difficult to understand Grafana integration. 

I'd like to know especially how to add, edit and persist changes in dashboard configuration, preferably with backups. I see two scenarios here, one is extending the default dashboards set by the cluster admin and the second one is persisting dashboards created by users (developers). The first scenario could be achieved via programatic way, the second is just about data persistence I guess. 

If you could point me how to add dashboards to default set and how to configure persistence in Grafana should be able to contribute some docs. 

Related: #98

**Comment by prov-thinrichs** [comment]
In our pipeline we use something similar to this:
```
kubectl patch configmap kube-prometheus-grafana --namespace monitoring --record=true -p "$(cat ./dashboards.yaml)" || true
```
So here's the breakdown:

``` kubectl patch configmap kube-prometheus-grafana``` 
All the default dashboards are in this configmap, but we want to add a dashboard, so we will patch the configmap

``` --namespace monitoring```
This is the namespace that we've decided to install monitoring stuff into

``` --record=true ```
Record the change in configmap

``` -p "$(cat ./dashboards.yaml)"  ```
Apply this patch to the configmap

```  || $true``` 
Patch will have a nonzero exit code if you want to patch something in that is already present and has the same byte-for-byte value.  IE kubectl patch is not idempotent

After this change the grafana-watcher will notice that the configmap has been changed and will reload the dashboards.  This can take a little while, long enough that if you are on the manage dashboard page in grafana, you can see the dashboards start showing up one or two at a time when you refresh the page

**Comment by VinceMD** [comment]
The name of the configmap has changed in 0.19

`kubectl patch configmap grafana-dashboard-definitions --namespace monitoring --record=true -p "$(cat ./dashboards.yaml)" || true`

Is there a doc that explains this process?

**Comment by VinceMD** [comment]
Last I tried

```
Last I tried 

```
hack/grafana-dashboards-configmap-generator/bin/grafana_dashboards_generate.sh 
# Starting execution of grafana_dashboards_generate on 2018-05-11-222951
# Configured size limit: 240000 bytes
# Grafana input dashboards and datasources will be read from: pathprometheus-operator-0.19.0/contrib/kube-prometheus/hack/grafana-dashboards-configmap-generator/templates/grafana-dashboards
# Grafana Dashboards ConfigMap will be created into file:
pathprometheus-operator-0.19.0/contrib/kube-prometheus/hack/grafana-dashboards-configmap-generator/output/grafana-dashboards-configMap-2018-05-11-222951.yaml
# Grafana Deployment manifest will be created into file:
pathprometheus-operator-0.19.0/contrib/kube-prometheus/hack/grafana-dashboards-configmap-generator/output/grafana-deployment-2018-05-11-222951.yaml

hack/grafana-dashboards-configmap-generator/bin/grafana_dashboards_generate.sh: line 344: to_process: unbound variablehack/grafana-dashboards-configmap-generator/bin/grafana_dashboards_generate.sh 
# Starting execution of grafana_dashboards_generate on 2018-05-11-222951
# Configured size limit: 240000 bytes
# Grafana input dashboards and datasources will be read from: pathprometheus-operator-0.19.0/contrib/kube-prometheus/hack/grafana-dashboards-configmap-generator/templates/grafana-dashboards
# Grafana Dashboards ConfigMap will be created into file:
pathprometheus-operator-0.19.0/contrib/kube-prometheus/hack/grafana-dashboards-configmap-generator/output/grafana-dashboards-configMap-2018-05-11-222951.yaml
# Grafana Deployment manifest will be created into file:
pathprometheus-operator-0.19.0/contrib/kube-prometheus/hack/grafana-dashboards-configmap-generator/output/grafana-deployment-2018-05-11-222951.yaml

hack/grafana-dashboards-configmap-generator/bin/grafana_dashboards_generate.sh: line 344: to_process: unbound variable
```
:( it's getting very hard to understand how to use Prometheus Operator and which doc is the source of truth or which command to use

Let me know if I can help testing something

**Comment by VinceMD** [comment]
@brancz no worries. If you could give me some pointer on how to do it with 0.19 (adding dashboards), would be great. And also how to monitor any namespace (there is an option but no idea where to change that option :D )

I use kube-aws and kube-aws has just made it easier to use Prometheus by opening all the required ports (etcd, controllers. workers).
What I would love is to help write a doc so that anyone who setup kube-aws can also easily setup Prometheus and Grafana. I'm happy to write docs.

Basically step by step what commands to type after setting up kube-aws to get Prometheus and Grafana including etcd monitoring and adding custom dashboards to the mix.

I have written a doc to myself which worked up to 0.17 but with the rewrite it seems nothing works and I can't understand the Readme :)

**Comment by Renol7** [comment]
Yes, I use kube-prometheus **v0.19.0** and **I can't modify or add custom dashboards** because this version uses the external repo _github.com/brancz/kubernetes-grafana_ (specific commit 5df0eaa17361ce77f4a0486cc89f44f43bd51829) to download Grafana dashboards and does not use "assets/grafana" folder dashboards as https://github.com/coreos/prometheus-operator/blob/master/contrib/kube-prometheus/docs/developing-alerts-and-dashboards.md says.

How should I do that? Should we release a new version (0.19.1) so that we can use custom Grafana dashboards?

**Comment by brancz** [comment]
I'm in the process of writing the docs on how to do this, stay tuned, I'll open a PR today or tomorrow.

**Comment by brancz** [comment]
We have just updated the docs for Grafana dashboards and Prometheus rules: https://github.com/coreos/prometheus-operator/blob/master/contrib/kube-prometheus/docs/developing-prometheus-rules-and-grafana-dashboards.md

Let us know if anything is unclear! :slightly_smiling_face: 

I'll close this issue, but feel free to continue the discussion or open a new issue if there is anything about the new docs that does not work!

**Comment by Renol7** [comment]
Great! The generation of grafana dashboards works for me 🙂

Unfortunately, I prefer to use release versions, so I have to wait to test it better... I will continue to use 0.19.0 version.

Maybe we should document options available in the Makefile.
For example, "compiling" section (https://github.com/coreos/prometheus-operator/tree/master/contrib/kube-prometheus#compiling) could be replaced by `make generate-raw` instead of explaining the script. 
It would be useful for beginners like me (I have to find this option haha).

**Comment by brancz** [comment]
> Maybe we should document options available in the Makefile.
> For example, "compiling" section (https://github.com/coreos/prometheus-> operator/tree/master/contrib/kube-prometheus#compiling) could be replaced by make generate-raw instead of explaining the script.
> It would be useful for beginners like me (I have to find this option haha).

Thanks a lot for the valuable feedback! Would you like to create a PR and add your tipps to the readme for future users? :slightly_smiling_face: 

> Unfortunately, I prefer to use release versions, so I have to wait to test it better... I will continue to use 0.19.0 version.

We're hard at work getting the next release out, but we're working on some major features so it's still going to take a few weeks.

**Comment by Renol7** [comment]
> Thanks a lot for the valuable feedback! Would you like to create a PR and add your tipps to the readme for future users? 🙂

Sure! I am going to create it! 
And I will document more sections when the new version is available (because I will have to use the new features).

> We're hard at work getting the next release out, but we're working on some major features so it's still going to take a few weeks.

No worries, I will stay tuned! 🙂 I am going to use the new version until that!
